services/entitlement/service/src/messaging/handlers/entitlement-authenticated-message-handler.ts

```diff
@@ -1,48 +1,65 @@
 } from '@axinom/mosaic-id-guard';
+import { MessagingSettings } from '@axinom/mosaic-message-bus-abstractions';
+import { Logger, MosaicError } from '@axinom/mosaic-service-common';
 import {
-  MessageHandler,
-  MessageInfo,
-  OnMessageMiddleware,
-} from '@axinom/mosaic-message-bus';
-import { Dict } from '@axinom/mosaic-service-common';
-import { SubscriptionConfig } from 'rascal';
+  TransactionalInboxMessageHandler,
+  TypedTransactionalMessage,
+} from '@axinom/mosaic-transactional-inbox-outbox';
+import { DatabaseClient } from 'pg-transactional-outbox';
 import { Config } from '../../common';
 
 /**
- * Guard a message handler by getting and verifying the JWT token. It changes
- * the `MessageInfo` parameter to `AuthenticatedMessageInfo` which contains the
- * JWT token subject.
- * In addition it checks that the subjects permission match the required ones.
+ * Custom channel message handler to verify channel permissions
  */
-export abstract class EntitlementAuthenticatedMessageHandler<
-  TContent,
-> extends MessageHandler<TContent> {
+export abstract class EntitlementAuthenticatedTransactionalMessageHandler<
+  T,
+> extends TransactionalInboxMessageHandler<T, Config> {
   constructor(
-    messagingKey: string,
-    protected readonly config: Config,
-    overrides?: SubscriptionConfig,
-    middleware: OnMessageMiddleware[] = [],
+    messagingSettings: MessagingSettings,
+    logger: Logger,
+    messageProducerService:
+      | 'ax-monetization-grants-service'
+      | 'ax-subscription-monetization-service',
+    config: Config,
   ) {
-    super(messagingKey, overrides, [
-      ...middleware,
-      authenticationMiddleware({
-        tenantId: config.tenantId,
-        environmentId: config.environmentId,
-        authEndpoint: config.idServiceAuthBaseUrl,
-      }),
-    ]);
-  }
+    const authWrapper = async <TMessage>(
+      message: TypedTransactionalMessage<TMessage>,
+      dbClient: DatabaseClient,
+    ): Promise<GuardedContext> => {
+      const token = message.metadata.authToken;
+      if (token === undefined) {
+        throw new MosaicError(IdGuardErrors.AccessTokenRequired);
+      }
+      const subject = await getAuthenticatedManagementSubject(token, {
+        tenantId: this.config.tenantId,
+        environmentId: this.config.environmentId,
+        authEndpoint: this.config.idServiceAuthBaseUrl,
+      });
 
-  protected getPgSettings(message: MessageInfo): Dict<string> {
-    return buildAuthPgSettings(this.getSubject(message), this.config.serviceId);
-  }
+      // Check that the message producer had permissions on that service
+      const subjectPermissions = subject.permissions?.[messageProducerService];
+      if (
+        subjectPermissions === undefined ||
+        !Array.isArray(subjectPermissions)
+      ) {
+        throw new MosaicError({
+          code: IdGuardErrors.Unauthorized.code,
+          message: `Permission check failed as the subject has no permissions for the Channel Service.`,
```

Comment:
`Channel Service` needs to be changed to something common or dynamic (e.g. serviceId-related)